zio-http/src/main/scala/zhttp/service/Handler.scala

```diff
@@ -16,40 +16,92 @@ private[zhttp] final case class Handler[R](
+    }
+  }
 
-        override def url: URL = URL.fromString(jReq.uri()).getOrElse(null)
+  override def channelRead(ctx: Ctx, msg: Any): Unit = {
 
-        override def headers: Headers = Headers.make(jReq.headers())
+    implicit val iCtx: ChannelHandlerContext = ctx
+    msg match {
+      case jReq: FullHttpRequest =>
+        jReq.touch("server.Handler-channelRead0")
+        unsafeRun(
+          jReq,
+          app,
+          new Request {
+            override def method: Method = Method.fromHttpMethod(jReq.method())
+
+            override def url: URL = URL.fromString(jReq.uri()).getOrElse(null)
+
+            override def headers: Headers = Headers.make(jReq.headers())
+
+            override def remoteAddress: Option[InetAddress] = {
+              ctx.channel().remoteAddress() match {
+                case m: InetSocketAddress => Some(m.getAddress)
+                case _                    => None
+              }
+            }
+
+            override def data: HttpData = HttpData.fromByteBuf(jReq.content())
+          },
+        )
+      case jReq: HttpRequest     =>
+        unsafeRun(
+          jReq,
+          app,
+          new Request {
+            override def data: HttpData = HttpData.Incoming(callback =>
+              ctx
+                .pipeline()
+                .addAfter(HTTP_REQUEST_HANDLER, HTTP_CONTENT_HANDLER, new RequestBodyHandler(callback)): Unit,
+            )
+
+            override def headers: Headers = Headers.make(jReq.headers())
+
+            override def method: Method = Method.fromHttpMethod(jReq.method())
+
+            override def remoteAddress: Option[InetAddress] = {
+              ctx.channel().remoteAddress() match {
+                case m: InetSocketAddress => Some(m.getAddress)
+                case _                    => None
+              }
+            }
+
+            override def url: URL = URL.fromString(jReq.uri()).getOrElse(null)
+          },
+        )
+
+      case msg: HttpContent => ctx.fireChannelRead(msg): Unit
+
+      case _ =>
+        ctx.writeAndFlush(new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.NOT_ACCEPTABLE)): Unit
 
-        override def remoteAddress: Option[InetAddress] = {
-          ctx.channel().remoteAddress() match {
-            case m: InetSocketAddress => Some(m.getAddress)
-            case _                    => None
-          }
-        }
+    }
 
-        override def data: HttpData = HttpData.fromByteBuf(jReq.content())
-      },
-    )
+  }
+  private def upgradeWebsocket(req: HttpRequest, response: Response)(implicit ctx: Ctx): Unit = {
+    req match {
+      case request: FullHttpRequest => self.upgradeToWebSocket(ctx, request, response)
+      case jReq: HttpRequest        =>
+        val fullRequest = new DefaultFullHttpRequest(jReq.protocolVersion(), jReq.method(), jReq.uri())
+        fullRequest.headers().setAll(jReq.headers())
+        self.upgradeToWebSocket(ctx, fullRequest, response)
```

Comment:
Can be further optimized using a special Request type. Please create a separate issue for this.

Reply:
```scala
final class EmptyFullHttpRequest(val request: HttpRequest) extends FullHttpRequest { self =>
    override def copy(): FullHttpRequest = self

    override def duplicate(): FullHttpRequest = self

    override def retainedDuplicate(): FullHttpRequest = self

    override def replace(content: ByteBuf): FullHttpRequest = self

    override def retain(increment: Int): FullHttpRequest = self

    override def retain(): FullHttpRequest = self

    override def touch(): FullHttpRequest = self

    override def touch(hint: Any): FullHttpRequest = self

    override def setProtocolVersion(version: HttpVersion): FullHttpRequest = self

    override def setMethod(method: HttpMethod): FullHttpRequest = self

    override def setUri(uri: String): FullHttpRequest = self

    override def getMethod: HttpMethod = request.method()

    override def method(): HttpMethod = request.method()

    override def getUri: String = request.uri()

    override def uri(): String = request.uri()

    override def trailingHeaders(): HttpHeaders = null

    override def decoderResult(): DecoderResult = request.decoderResult()

    override def setDecoderResult(result: DecoderResult): Unit = ()

    override def refCnt(): Int = 0

    override def release(): Boolean = false

    override def release(decrement: Int): Boolean = false

    override def getDecoderResult: DecoderResult = request.getDecoderResult()

    override def content(): ByteBuf = ???

    override def getProtocolVersion: HttpVersion = request.protocolVersion()

    override def protocolVersion(): HttpVersion = request.protocolVersion()

    override def headers(): HttpHeaders = request.headers()
  }
```

Something like this ☝️